common/thorhelper/thorread.cpp

```diff
@@ -0,0 +1,825 @@
+    IMPLEMENT_IINTERFACE
+
+    virtual IRawRowStream * queryRawRowStream() override;
+    virtual IAllocRowStream * queryAllocatedRowStream(IEngineRowAllocator * _outputAllocator) override;
+
+    virtual bool getCursor(MemoryBuffer & cursor) override;
+    virtual void setCursor(MemoryBuffer & cursor) override;
+    virtual void stop() override;
+
+    virtual void clearInput() override;
+    virtual bool matches(const char * format, bool streamRemote, unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options) override;
+
+// IThorDiskCallback
+    virtual offset_t getFilePosition(const void * row) override;
+    virtual offset_t getLocalFilePosition(const void * row) override;
+    virtual const char * queryLogicalFilename(const void * row) override;
+    virtual const byte * lookupBlob(unsigned __int64 id) override { UNIMPLEMENTED; }
+
+
+protected:
+    offset_t getLocalOffset();
+
+protected:
+    Owned<ISerialStream> input;
+    Owned<IFileIO> inputfileio;
+    CThorContiguousRowBuffer inputBuffer;
+    Owned<IEngineRowAllocator> outputAllocator;
+    IOutputMetaData * expectedDiskMeta = nullptr;
+    IOutputMetaData * projectedDiskMeta = nullptr;
+    IOutputMetaData * actualDiskMeta = nullptr;
+    unsigned expectedCrc;
+    unsigned projectedCrc;
+    unsigned actualCrc;
+    MemoryBuffer encryptionKey;
+    size32_t readBufferSize = defaultReadBufferSize;
+    bool grouped = false;
+    bool stranded = false;
+    bool compressed = false;
+    bool blockcompressed = false;
+    bool rowcompressed = false;
+
+//The following refer to the current input file:
+    offset_t fileBaseOffset = 0;
+    StringAttr logicalFilename;
+    unsigned filePart = 0;
+};
+
+
+DiskRowReader::DiskRowReader(unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options)
+: expectedCrc(_expectedCrc), expectedDiskMeta(&_expected), projectedCrc(_projectedCrc), projectedDiskMeta(&_projected), actualCrc(_actualCrc), actualDiskMeta(&_actual)
+{
+    if (options)
+    {
+        //Options contain information that is the same for each file that is being read, and potentially expensive to reconfigure.
+        if (options->hasProp("encryptionKey"))
+        {
+            encryptionKey.resetBuffer();
+            options->getPropBin("encryptionKey", encryptionKey);
+        }
+        readBufferSize = options->getPropInt("readBufferSize", defaultReadBufferSize);
+    }
+}
+
+IRawRowStream * DiskRowReader::queryRawRowStream()
+{
+    return this;
+}
+
+IAllocRowStream * DiskRowReader::queryAllocatedRowStream(IEngineRowAllocator * _outputAllocator)
+{
+    outputAllocator.set(_outputAllocator);
```

Comment:
feels like it's doing more than method name suggests (queryAllocatedRowStream()).
Should row allocate be set via a different method instead?